sys/net/application_layer/coapfileserver/coapfileserver.c

```diff
@@ -0,0 +1,426 @@
+     * If we did a sharper estimation (factoring in the block2 size option with
+     * the current blockum), we'd even pack 512 bytes into 530 until block
+     * numbers get large enough to eat another byte, which is when the block
+     * size would decrease in-flight. */
+    size_t remaining_length = len - (pdu->payload - buf);
+    remaining_length -= 5; /* maximum block2 option usable in nanocoap */
+    remaining_length -= 1; /* payload marker */
+    /* > 0: To not wrap around; if that still won't fit that's later caught in
+     * an assertion */
+    while ((coap_szx2size(request->szx2) > remaining_length) && (request->szx2 > 0)) {
+        request->szx2--;
+        request->blocknum2 <<= 1;
+    }
+    coap_block_slicer_t slicer;
+    coap_block_slicer_init(&slicer, request->blocknum2, coap_szx2size(request->szx2));
+    coap_opt_add_block2(pdu, &slicer, true);
+    size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+    err = vfs_lseek(fd, slicer.start, SEEK_SET);
+    if (err < 0) {
+        goto late_err;
+    }
+
+    /* That'd only happen if the buffer is too small for even a 16-byte block,
+     * or if the above calculations were wrong.
+     *
+     * Not using payload_len here as that's needlessly underestimating the
+     * space by CONFIG_GCOAP_RESP_OPTIONS_BUF
+     * */
+    assert(pdu->payload + slicer.end - slicer.start <= buf + len);
+    int read = vfs_read(fd, pdu->payload, slicer.end - slicer.start);
+    if (read < 0) {
+        goto late_err;
+    }
+
+    uint8_t morebuf;
+    int more = vfs_read(fd, &morebuf, 1);
+    if (more < 0) {
+        goto late_err;
+    }
+
+    vfs_close(fd);
+
+    slicer.cur = slicer.end + more;
+    coap_block2_finish(&slicer);
+
+    if (read == 0) {
+        /* Rewind to clear payload marker */
+        read -= 1;
+    }
+
+    return resp_len + read;
+
+late_err:
+    vfs_close(fd);
+    coap_hdr_set_code(pdu->hdr, COAP_CODE_INTERNAL_SERVER_ERROR);
+    return coap_get_total_hdr_len(pdu);
+}
+
+static ssize_t coapfileserver_directory_handler(coap_pkt_t *pdu, uint8_t *buf, size_t len, struct requestdata *request)
+{
+    /**
+     * ToDo:
+     *
+     * * Blockwise
+     *
+     */
+    vfs_DIR dir;
+
+    int err = vfs_opendir(&dir, request->namebuf);
+    if (err != 0) {
+        return coapfileserver_errno_handler(pdu, buf, len, err);
+    }
+    DEBUG("coapfileserver: Serving directory listing\n");
+
+    gcoap_resp_init(pdu, buf, len, COAP_CODE_CONTENT);
+    coap_opt_add_format(pdu, COAP_FORMAT_LINK);
+    size_t resp_len = coap_opt_finish(pdu, COAP_OPT_FINISH_PAYLOAD);
+
+    vfs_dirent_t entry;
+    ssize_t payload_cursor = 0;
+    while (vfs_readdir(&dir, &entry) > 0) {
+        char *entry_name = entry.d_name;
+        /* Ignore leading slashes presented by some file systems; see
+         * https://github.com/RIOT-OS/RIOT/issues/14635 */
+        while (entry_name[0] == '/') {
+            entry_name += 1;
```

Comment:
I think we can drop this now, but what's confusion me: `aiocoap-fileserver` seems to always include the leading slash:

#### coapfileserver

```
% aiocoap-client coap://[2001:16b8:4585:2800:fec2:3dff:fe23:22df]/vfs/
application/link-format content was re-formatted
<song.txt>,
<test/>
```

#### aiocoap-fileserver

```
% aiocoap-client coap://[::1]/
Invalid application/link-format content was not re-formatted
</<foo>>,</song.txt>,</test/>;ct=40
```

Reply:
The leading slash is not mandatory -- thing is: <song.txt> is, following the rules of RFC6690 (note that "Origin", in "Origin of", is not capitalized as the start of an item, but because it is a proper term) in `/vfs/` does not refer to `/vfs/song.txt`, as one might assume, but to `/song.text`.

aiocoap-fileserver thus prints the slash to make it explicit, and to not give false hopes to people that they could append the path. It should also output the full path (because relative paths just don't work).

This implementation will need to do the same, or use a different format than link-format (HTML would work fine ;-) ).

Also, if the 'foo' is from a file literally named "`<foo>`": 6690 just can't do them.

Reply:
> This implementation will need to do the same

+1
So always printing the full path in directory listings (with leading slash), like

```
% aiocoap-client coap://[::1]/test/            :(
application/link-format content was re-formatted
</test/bar/>; ct=40,
</test/test.txt>
```

> or use a different format than link-format (HTML would work fine ;-) ).

Since my plan is to write a client that works with *both* `aiocoap-fileserver` and `coapfileserver` (maybe I rename it to `gcoap_fileserver` to avoid confusion) that will not be ideal :wink:

Reply:
aiocoap-fileserver will likely start supporting different directory index formats -- 6690 link format would stay supported, but it may need to be explicitly Accept-ed. I guess the default will become CoRAL at some point. Support for HTML or anything other that seems useful would be easily possible.

Reply:
(The Rust implementation currently avoids the issue altogether by serving text/plain index files...)